lib/stud/interval.rb

```diff
@@ -9,30 +12,87 @@ module Stud
         block.call
       else
         block.call
         start = sleep_for_interval(time, start)
+        break if stop?
       end
     end # loop forever
   end # def interval
 
+  def interval(time, opts = {}, &block)
+    Stud.interval(time, opts, &block)
+  end # def interval
+
+  # stop! instructs interval to stop and exit its execution loop before going to
+  # sleep between block executions.
+  # NOW the tricky part is: this is typically an operation that will be called
+  # from another thread than the thread running the interval loop in which case
+  # the target parameter must be set to the Thread object which is running the
+  # interval loop.
+  # Note that the stop logic is compatible with Stud::Task so if interval is run
+  # inside a Stud::Task, calling Stud::Task#stop! will stop the interval the same
+  # way as calling stop! on the interval itself.
+  # @param target [Thread] the target thread to stop, defaut to Thread.current
+  def self.stop!(target = Thread.current)
+    # setting/getting threalocal var is thread safe in JRuby
+    target[STUD_STOP_REQUESTED] = true
+    target.wakeup
+    nil
+  end
+
+  # stop? returns true if stop! has been called
+  # @param target [Thread] the target thread to check for stop, defaut to Thread.current
+  # @return [Boolean] true if the stop! has been called
+  def self.stop?(target = Thread.current)
+    # setting/getting threalocal var is thread safe in JRuby
+    target[STUD_STOP_REQUESTED]
+  end
+
+  class << Stud
+    # also support Stud.interrupted? for backward compatibility.
+    alias_method :interrupted?, :stop?
+  end
+
+  # stoppable_sleep will try to sleep for the given duration seconds (which may be any number,
+  # including a Float with fractional seconds). an optional stop_condition_block can be supplied
+  # to verify for sleep interruption if the block returns a truthy value. if not block is supplied
+  # it will check for the Stud.stop? condition. this check will be performed at 1s interval
+  # by default or you can supply a different stop_condition_interval.
+  # @param duration [Numeric] sleep time in (fractional) seconds
+  # @param stop_condition_interval [Numeric] optional interval in (fractional) seconds to perform the sleep interruption verification, default is 1s
+  # @param stop_condition_block [Proc] optional sleep interruption code block that must evaluate to a truthy value, default is to use Stud.stop?
+  def self.stoppable_sleep(duration, stop_condition_interval = 1.0, &stop_condition_block)
+    sleep_start = Time.now
+
+    while (remaining_duration = (duration - (Time.now - sleep_start))) >= stop_condition_interval
+      # sleep X if there is more than X remaining to sleep in relation to the loop start time
+      sleep(stop_condition_interval)
+
+      return if block_given? ? stop_condition_block.call : stop?
+    end
+
+    # here we know we have less than 1s reminding to sleep,
+    sleep(remaining_duration) if remaining_duration > 0.0
```

Comment:
My reasoning below was more of a convenience, I don't believe we need to follow the sleep contract, since we can interrupt ours.. :) 

I agree with @colinsurprenant It could be handy to return both duration and `stopped?` the predicate allow us to not keep track of the time we use as the argument for our `stoppable_sleep`